src/OpenLoco/src/Windows/ObjectSelectionWindow.cpp

```diff
@@ -517,128 +505,116 @@ namespace OpenLoco::Ui::Windows::ObjectSelectionWindow
             self.widgets[widx::closeButton].type = WidgetType::none;
         }
 
-        // Update the title.
+        const auto& currentTab = kMainTabInfo[self.currentTab];
+        const auto& subTabs = currentTab.subTabs;
+        const bool showSecondaryTabs = !subTabs.empty() && FilterLevel(self.var_856) != FilterLevel::beginner;
+
+        // Update page title
         auto args = FormatArguments(self.widgets[widx::caption].textArgs);
-        args.push(_tabDisplayInfo[self.currentTab].name);
+        if (showSecondaryTabs)
+            args.push(subTabs[self.currentSecondaryTab].name);
+        else
+            args.push(kMainTabInfo[self.currentTab].name);
 
-        const auto& tabFlags = _tabDisplayInfo[self.currentTab].flags;
-        const bool showSecondaryTabs = (tabFlags & ObjectTabFlags::filterByVehicleType) != ObjectTabFlags::none;
-        for (int i = widx::vehicleTypeTrain; i <= widx::vehicleTypeShip; i++)
+        // Toggle secondary tabs
+        for (auto i = 0U; i < 8U; i++)
         {
-            self.widgets[i].type = showSecondaryTabs ? WidgetType::tab : WidgetType::none;
+            const auto widgetIndex = i + widx::secondaryTab1;
+
+            const bool subTabIsVisible = showSecondaryTabs && i < subTabs.size() && shouldShowSubTab(subTabs, i, FilterLevel(self.var_856));
+            if (!subTabIsVisible)
+                self.disabledWidgets |= 1ULL << widgetIndex;
```

Comment:
we really need some helper functions for this bit-toggling, its everywhere

Reply:
Yes, we already have an `Window::isDisabled(WidgetIndex_t)`. We could add a `setDisabled` and `unsetDisabled`, etc. Bit out of scope for this PR, though.